src/Microsoft.Health.Fhir.Shared.Api/Controllers/ImportController.cs

```diff
@@ -79,9 +93,80 @@ public class ImportController : Controller
             _features = features.Value;
             _mediator = mediator;
+            _resourceWrapperFactory = resourceWrapperFactory;
             _logger = logger;
         }
 
+        [HttpPost]
+        [Route(KnownRoutes.ImportBundle)]
+        [AuditEventType(AuditEventSubType.ImportBundle)]
+        public async Task<IActionResult> ImportBundle()
+        {
+            var sw = Stopwatch.StartNew();
+            var startDate = DateTime.UtcNow;
+            var resources = new List<ImportResource>();
+            var keys = new HashSet<ResourceKey>();
+            var fhirParser = new FhirJsonParser();
+            var importParser = new ImportResourceParser(fhirParser, _resourceWrapperFactory);
+            var index = 0L;
+            using var reader = new StreamReader(Request.Body, Encoding.UTF8);
+            try
+            {
+                if (Request.ContentType != null && Request.ContentType.Contains("application/fhir+ndjson", StringComparison.OrdinalIgnoreCase))
+                {
+                    var line = await reader.ReadLineAsync();
+                    while (line != null)
+                    {
+                        var importResource = importParser.Parse(index, 0, 0, line, ImportMode.IncrementalLoad);
+                        var key = importResource.ResourceWrapper.ToResourceKey(true);
+                        if (!keys.Add(key))
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.ResourcesMustBeUnique, key));
+                        }
+
+                        resources.Add(importResource);
+                        line = await reader.ReadLineAsync();
+                        index++;
+                    }
+                }
+                else
+                {
+                    var str = await reader.ReadToEndAsync();
+                    var resource = await fhirParser.ParseAsync(str);
+                    var bundle = resource.ToResourceElement().ToPoco<Bundle>();
+                    foreach (var entry in bundle.Entry)
+                    {
+                        if (entry.Request?.Method != Bundle.HTTPVerb.PUT)
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.InvalidBundleEntry, entry.Request?.Method, KnownRoutes.ImportBundle));
+                        }
+
+                        var importResource = importParser.Parse(index, 0, 0, entry.Resource, ImportMode.IncrementalLoad);
+                        var key = importResource.ResourceWrapper.ToResourceKey(true);
+                        if (!keys.Add(key))
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.ResourcesMustBeUnique, key));
+                        }
+
+                        resources.Add(importResource);
+                        index++;
+                    }
+                }
+            }
+            catch (Exception)
```

Comment:
This exception is too broad. 
If any type of exception is thrown, then it'll return a "Unable to parse resource at index=", given that it's not necessarily and parsing issue.

Reply:
Per example, when a RequestNotValidException is thrown with a ResourcesMustBeUnique or an InvalidBundleEntry, the details of those exceptions are lost.

Reply:
What exception type do you suggest?

Reply:
I changed the exception. Please keep in mind that current logic is "fail fast", it exits on first found problem. This is intentional.